### PR TITLE
[RFC] ASoC: SOF: Intel-HDA - fix capture issue on WHL&GLK

### PR DESCRIPTION
**Pull request by RanderWang**
issue: The pcm data is full of zero when capturing on WHL&GLK.
    
    There is a group of stream registers in HDA controller
    to control a set of dma channels in dsp, and one stream
    register is bind to one dma channel. Driver set each stream
    register to control the behavior of corresponding dma channel.
    The issue is caused by mismatch between host stream register
    and link dma channel in FW. The resultis that nothing is gotten
    by link dma
    
    There are two dma modes in dsp:
    (1) host dma and link dma work in decouple mode for SOF + HDA
    (2) host dma and GP dma work together for SOF + I2S
    In decopule mode, the host dma is set by dai frontend and link
    dma is set by dai backend and the setting maybe different.
    In SOF + I2S mode, host dma and GP dma are also independent to
    each other and GP dma channel is allocated in FW instead of in host
    
    Driver doesn't support mode(1) and only sends the setting of host
    dma to control both host dma and link dma, so this causes the mismatch
    
    To make host driver set correct register
    (1) the dma channel is allocated in host and sent to FW to allocate
        dma channel and host set the stream register according to this dma
        channel.
    (2) driver sends both host dma channel and BE dma (link dma)
        channel to FW

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

**Comment by RanderWang** [comment]
FW change: https://github.com/thesofproject/sof/pull/516
Liam suggested that: hda is special to intel and not generic thing, so I just put the HDA code in intel domain

No change is for BYT or HSW because the dma channel is allocated by FW.

This patch supports for multiple dai backends. In SOF, each component
has a unique id which is useful to my patch. Now host stores all the
dma setting of each BE and its id, then sends them to FW to set each BE.
In FW, each BE could get dma setting by its id

**Comment by plbossart** [comment]
I don't think it makes sense to assume link DMA chan ID and host DMA chan ID have to be identical. This will not work with 1:N or N:1 configurations.

**Comment by plbossart** [comment]
@xiulipan @mengdonglin can you look into the CI status. I get both a build fails and a log that tells me the build it go.

File: include/uapi/sound/sof-ipc.h [hunk 1]
```diff
@@ -476,6 +476,7 @@ struct sof_ipc_stream_params {
 	/* for notifying host period has completed - 0 means no period IRQ */
 	uint32_t host_period_bytes;
 	enum sof_ipc_chmap chmap[SOF_IPC_MAX_CHANNELS];	/* channel map */
+	uint32_t private_data[64];
```
Comment (plbossart) [inline include/uapi/sound/sof-ipc.h]: Did you mean 'reserved'. Private data means you are using it in a opaque way, reserved means don't mess with this structure, it's for future usages.

File: sound/soc/sof/core.c [hunk 2]
```diff
@@ -96,7 +96,7 @@ struct snd_sof_pcm *snd_sof_find_spcm_pcm_id(struct snd_sof_dev *sdev,
 }
 
 struct snd_sof_widget *snd_sof_find_swidget(struct snd_sof_dev *sdev,
-					    char *name)
+					    const char *name)
 {
```
Comment (plbossart) [inline sound/soc/sof/core.c:100]: this is a different patch, send it in a different PR

File: sound/soc/sof/intel/hda-pcm.c [hunk 3]
```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
...
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work in couple mode for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in couple mode, host dma and GP dma share the same
+ * dma setting.
+ */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c:105]: No. you are using the term 'coupled' mode is an overloaded way. In the I2S mode, the Host and GP DMA have nothing to do with each other, and likewise in HDaudio mode there should be no dependency.
I don't see why you introduce new confusing concepts

File: sound/soc/sof/intel/hda-dai.c [hunk 4]
```diff
@@ -158,6 +155,7 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
 	if (!link)
 		return -EINVAL;
 
+	snd_soc_dai_set_dma_data(dai, substream, NULL);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:158]: what exactly does this do here? What information are you trying to save?

File: sound/soc/sof/intel/hda-pcm.c [hunk 5]
```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
...
+				     struct snd_pcm_substream *substream,
+				     struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *be_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_intel_be_params *be_params;
+
+	be_params = (struct sof_intel_be_params *)ipc_params->private_data;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c]: why do you need private data here?
I am confused on what you are trying to do.

File: sound/soc/sof/intel/hda-pcm.c [hunk 6]
```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
...
+			/* link dma channel is derived from stream_tag
+			 * in decouple mode
+			 */
+			be_params->be_dma_ch[index] =
+					be_stream->hstream.stream_tag - 1;
+		} else {
+			/* the setting of GP dma is the same as host dma
+			 * in default. Actually, this value is ignored by
+			 * allocation function for GP dma
+			 */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c]: Again this makes no sense, the settings of GP DMA and host DMA are completely independent. The former are not known to the driver.
Also why do you need to set a value if it's ignored?

File: sound/soc/sof/intel/hda-pcm.c [hunk 7]
```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
...
+			dapm_widget = be->cpu_dai->capture_widget;
+
+		sof_widget = snd_sof_find_swidget(sdev, dapm_widget->name);
+		if (!sof_widget) {
+			dev_err(sdev->dev, "error: failed to find backend widget");
+			return -EINVAL;
+		}
+
+		be_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
+		if (be_stream) {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c]: Can you clarify what this test means.

**Comment by lgirdwood** [comment]
@plbossart yeah, I still dont know the rules for mapping @RanderWang it's also not proven why the driver needs to do this atm.....   @RanderWang  I think given a stream.id from the host then the platform/cavs logic in the FW should then be able to select the correct DMA link ID without any further host intervention ?

**Comment by RanderWang** [comment]
> @plbossart yeah, I still dont know the rules for mapping @RanderWang it's also not proven why the driver needs to do this atm..... @RanderWang I think given a stream.id from the host then the platform/cavs logic in the FW should then be able to select the correct DMA link ID without any further host intervention ?

link dma

here each link dma channel is bind to a PPLCxCTL register in HDA controller(HW link, cant be configured). Host sets PPLCxCTL to set the target of corresponding link dma(PPLC0CTL controls link dma channel 0,  n <->n). Each HDA codec
has a id and only get the data package with this id. Link dma copys data to HDA hub which connect to all the HDA codecs. PPLCxCTL set the id to the data stream of corresponding link dma channel and route it to the HDA codec with the id

PPLCxCTL is set by host and link dma channel is allocated by FW in DSP.
Now the algorithm is, allocate channel in host and set corresponding PPLCxCTL regsiter then send the channel to FW to allocate

Link dma is independent of host dma

**Comment by xiulipan** [comment]
@plbossart 
Will enable Travis CI as backup ASAP.
The inline checkpatch may need time to check and test which plugin in github can do the job.

File: sound/soc/sof/intel/hda-pcm.c (hunk 7 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c]: The be_stream is stored by HDA dai backend. For I2S backend, there is no such operation.
So check it to make clear the backend is HDA or I2S

File: sound/soc/sof/intel/hda-pcm.c (hunk 6 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c]: yes, I set it just to initialize a default value. It may be safe to reduce some bugs.
I will remove it.

File: sound/soc/sof/intel/hda-pcm.c (hunk 5 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c]: Liam suggested that: hda is special to intel and not generic thing. If sony or other vendors want to use SOF, they don't like link_dma_channel, and they may also have some special setting to send to FW. I discussed with team members about this, and decide to prepare a private_data for this case.

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:158]: Here is just to clear saving. The dma data is saved at hda_link_hw_params, and cleared at hda_link_hw_free

File: sound/soc/sof/intel/hda-pcm.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c:105]: yes, it is exactly overloaded. I just want to make it clear but now confusing, I will refine it

File: include/uapi/sound/sof-ipc.h (hunk 1 above)
Comment (RanderWang) [reply include/uapi/sound/sof-ipc.h]: it is private for HDA.  Please see comment below

File: sound/soc/sof/core.c (hunk 2 above)
Comment (RanderWang) [reply sound/soc/sof/core.c:100]: ok, I will refine it

**Comment by RanderWang** [comment]
I change the definition of param to following one, but I get a bug and I am debugging it now.

struct sof_ipc_stream_params {
	struct sof_ipc_host_buffer buffer;
	enum sof_ipc_stream_direction direction;
	enum sof_ipc_frame frame_fmt;
	enum sof_ipc_buffer_format buffer_fmt;
	uint32_t stream_tag;
	uint32_t rate;
	uint32_t channels;
	uint32_t sample_valid_bytes;
	uint32_t sample_container_bytes;
	/* for notifying host period has completed - 0 means no period IRQ */
	uint32_t host_period_bytes;
	enum sof_ipc_chmap chmap[SOF_IPC_MAX_CHANNELS];	/* channel map */
	uint32_t private_size;
	uint32_t private_data[0];
} __attribute__((packed));

File: include/uapi/sound/sof-ipc.h (hunk 1 above)
Comment (plbossart) [reply include/uapi/sound/sof-ipc.h]: it's ugly, there is no reason to make things opaque. better use a union as done for the SSP/dmic.
Also you are not getting the point. 'reserved' means not currently used but may be reclaimed in future versions. It's completely different from private or HDAudio specific where you do make use of this. You do not need 64 bytes to handle HDaudio, but it would make sense to add 64 bytes of reserved data to be future-proof.

File: sound/soc/sof/intel/hda-pcm.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c]: use a union instead if you are concerned about code size or non-common parameters.

**Comment by RanderWang** [comment]
update my PR, thanks for review

File: sound/soc/sof/intel/hda-pcm.c (hunk 3 above)
Comment (lgirdwood) [reply sound/soc/sof/intel/hda-pcm.c:105]: yeah, "coupled" to me means there is a dependency relationship between the two DMACs meaning you have the meaning incorrect for 1 & 2 above. i.e. there is no coupling with between Host GW and GP DMA. 

Can you tell name the DSP registers that are used in binding, it's not obvious from the HAS.

File: include/uapi/sound/sof-ipc.h [hunk 8]
```diff
@@ -278,7 +284,9 @@ struct sof_ipc_dai_ssp_params {
 /* HDA Configuration Request - SOF_IPC_DAI_HDA_CONFIG */
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
-	/* TODO */
+	u32 hda_count;
+	u32 hda_comp_id[SOF_IPC_MAX_CHANNELS];
+	u32 hda_dma_ch[SOF_IPC_MAX_CHANNELS];
 } __attribute__((packed));
```
Comment (plbossart) [inline include/uapi/sound/sof-ipc.h:290]: Sorry, I don't get why this SOF_IPC_MAX_CHANNELS constant is used here? This is used elsewhere for channel maps, which has nothing to do with HDAudio components or DMA?
Is a new definition needed and why?

Comment (plbossart) [inline include/uapi/sound/sof-ipc.h:290]: what is the use of this hda_params structure?
I really don't understand how this might work long term, e.g if we have two host-DSP transfers (low-latency and  media) which get mixed on the same output? Are we going to send the same hda_dma_ch values twice, possibly after the link DMA is already running to deal with the first stream?
The problem for me is that you stick the hdaudio params in the IPC stream params when they should be in this

````text 
 /* HDA Configuration Request - SOF_IPC_DAI_HDA_CONFIG */
struct sof_ipc_dai_hda_params {
	struct sof_ipc_hdr hdr;
	/* TODO */
} __attribute__((packed));
````
Same if we add support for tone generation, you would need to pass the link DMA config *WITHOUT* any ipc for streams

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:158]: does this below in the same patch? what is the link with the DMA link configuration, it looks like a miss in previous patches rather?

File: sound/soc/sof/intel/hda-pcm.c [hunk 9]
```diff
@@ -95,6 +95,78 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
...
+	struct hdac_ext_stream *hda_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_ipc_dai_hda_params *hda_params;
+
+	hda_params = &ipc_params->hda;
+	hda_params->hda_count = 0;
+
+	/* travel BE clients to get BE stream with link dma id */
+	list_for_each_entry(dpcm, &fe->dpcm[direction].be_clients, list_be) {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c:120]: Sorry, I really don't get this loop either.
What happens when you have a mix of Hdaudio and SSP/DMIC? Why would you loop on all BE clients, there should be something that checks if this is relevant?
Also as I mentioned it earlier in the review, don't you have cases where the information will be configured multiple times, or none?
I know we sound painful but essentially your design remains obscure, sorry.

File: sound/soc/sof/intel/hda-pcm.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c:105]: yes, it is not obvious from the HAS. I just did some tests to get it.
In programming guide, the register is: PPLCxCTL

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:158]: yes, it is missed by previous patches. I just check this data and found it should be cleared.

File: include/uapi/sound/sof-ipc.h (hunk 8 above)
Comment (RanderWang) [reply include/uapi/sound/sof-ipc.h:290]: ok, I will define a new one.

File: sound/soc/sof/intel/hda-pcm.c (hunk 9 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c:120]: yes, actually there is a check in loop. Now I don't find a way to  check it is SSP or DMIC.

hda_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
if (hda_stream) {
...
here, hda_stream is only set by HDA dai backend. For other dai backend, it should be NULL.
And in FW, only HDA link dma use channel info.

For multiple dai backends, I need to check all the components of pipeline, form kernel, topology to Fw. We never test this case, so I need to spend some time to make it ready.

File: include/uapi/sound/sof-ipc.h (hunk 8 above)
Comment (RanderWang) [reply include/uapi/sound/sof-ipc.h:290]: Hi Pierre, your questions are discussed many times and no conclusion is gotten.
For question one, @lgirdwood @xiulipan @tlauda can you help to make it clear.
For tone, @lgirdwood  @ranj063 , can you help to make it clear.
Thanks!

File: sound/soc/sof/intel/hda-pcm.c (hunk 9 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-pcm.c:120]: I don't feel painful about implementing a feature. We have a plan with CCG team, so they often ask me which branch to use. This would trouble me a little.

Comment (lgirdwood) [reply sound/soc/sof/intel/hda-pcm.c:120]: @RanderWang it would help if you describe the host registers used to get the channel info and the DSP link DMA registers you program the data into. That we we can read the docs and understand the problem better.

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:158]: ok so can you refactor this out and submit a different PR with all the changes that shouldn't be here? I have no problem merging this, it's unrelated to the link DMA stuff.

File: sound/soc/sof/intel/hda-pcm.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-pcm.c:120]: I am with @lgirdwood, we don't quite understand what registers need to be programmed and when. Specifically I don't see why there is a coupling between host and link DMA in terms of IPC when precisely we operate in de-coupled mode...
